File: src/commands/grades.ts
import { Command } from "commander";
import type { OutputFormat } from "../lib/types.ts";
import { getEnrolledCoursesApi, getCourseGradesApi } from "../lib/moodle.ts";
import { createApiContext } from "../lib/auth.ts";
import { formatAndOutput } from "../index.ts";
import { getOutputFormat } from "../lib/utils.ts";

interface GradeSummary {
  courseId: number;
  courseName: string;
  grade?: string;
  gradeFormatted?: string;
  rank?: number;
  totalUsers?: number;
}

export function registerGradesCommand(program: Command): void {
  const gradesCmd = program.command("grades");
  gradesCmd.description("Grade operations");

  gradesCmd
    .command("summary")
    .description("Show grade summary across all courses")
    .option("--output <format>", "Output format: json|csv|table|silent")
    .action(async (options, command) => {
      const output: OutputFormat = getOutputFormat(command);
      const apiContext = await createApiContext(options, command);
      if (!apiContext) {
        process.exitCode = 1;
        return;
      }

      const courses = await getEnrolledCoursesApi(apiContext.session);

      const gradeResults = await Promise.allSettled(
        courses.map(course =>
          getCourseGradesApi(apiContext.session, course.id)
            .then(grades => ({ course, grades }))
        )
      );

      const gradeSummaries: GradeSummary[] = [];
      for (const result of gradeResults) {
        if (result.status !== "fulfilled") continue;
        const { course, grades } = result.value;
        gradeSummaries.push({
          courseId: course.id,
          courseName: course.fullname,
          grade: grades.grade,
          gradeFormatted: grades.gradeFormatted,
          rank: grades.rank,
          totalUsers: grades.totalUsers,
        });
      }

      const gradedCourses = gradeSummaries.filter(g => g.grade !== undefined && g.grade !== null && g.grade !== "-");
      const averageRank = gradeSummaries
        .filter(g => g.rank !== undefined && g.rank !== null)
        .reduce((sum, g) => sum + (g.rank || 0), 0) /
        (gradeSummaries.filter(g => g.rank !== undefined && g.rank !== null).length || 1);

      apiContext.log.info(`Total: ${courses.length} courses, ${gradedCourses.length} graded, avg rank: ${averageRank.toFixed(1)}`);

      formatAndOutput(gradeSummaries as unknown as Record<string, unknown>[], output, apiContext.log);
    });

  gradesCmd
    .command("course")
    .description("Show detailed grades for a specific course")
    .argument("<course-id>", "Course ID")
    .option("--output <format>", "Output format: json|csv|table|silent")
    .action(async (courseId, options, command) => {
      const output: OutputFormat = getOutputFormat(command);
      const apiContext = await createApiContext(options, command);
      if (!apiContext) {
        process.exitCode = 1;
        return;
      }

      const courses = await getEnrolledCoursesApi(apiContext.session);
      const course = courses.find(c => c.id === parseInt(courseId, 10));

      if (!course) {
        apiContext.log.error(`Course not found: ${courseId}`);
        process.exitCode = 1;
        return;
      }

      const grades = await getCourseGradesApi(apiContext.session, course.id);

      const gradeData = {
        courseId: grades.courseId,
        courseName: grades.courseName,
        grade: grades.grade,
        gradeFormatted: grades.gradeFormatted,
        rank: grades.rank,
        totalUsers: grades.totalUsers,
        items: grades.items?.map(item => ({
          name: item.name,
          grade: item.grade,
          gradeFormatted: item.gradeFormatted,
          range: item.range,
          percentage: item.percentage,
          weight: item.weight,
          feedback: item.feedback,
          graded: item.graded,
        })),
      };

      formatAndOutput(gradeData as unknown as Record<string, unknown>, output, apiContext.log);
    });
}

File: src/commands/grades.rs
use anyhow::Result;
use crate::Cli;
use crate::moodle::course::get_enrolled_courses_api;
use crate::moodle::grade::get_course_grades_api;
use crate::output::format_and_output;
use super::ApiCtx;

pub async fn run(cmd: &crate::GradesCommands, cli: &Cli) -> Result<()> {
    let ctx = ApiCtx::build(cli)?;

    match cmd {
        crate::GradesCommands::Summary => {
            let courses = get_enrolled_courses_api(&ctx.client, &ctx.session, "inprogress").await?;

            let mut summaries: Vec<serde_json::Value> = Vec::new();

            for course in &courses {
                match get_course_grades_api(&ctx.client, &ctx.session, course.id, ctx.session.user_id).await {
                    Ok(mut grade) => {
                        grade.course_name = course.fullname.clone();
                        summaries.push(serde_json::json!({
                            "courseId": grade.course_id,
                            "courseName": grade.course_name,
                            "grade": grade.grade,
                        }));
                    }
                    Err(e) => {
                        ctx.log.warn(&format!("Failed to get grades for course {}: {}", course.id, e));
                    }
                }
            }

            let graded = summaries.iter().filter(|g| {
                g.get("grade").map(|v| !v.is_null() && v.as_str() != Some("-")).unwrap_or(false)
            }).count();

            ctx.log.info(&format!(
                "Total: {} courses, {} graded",
                courses.len(), graded,
            ));

            format_and_output(&summaries, ctx.output, None);
        }

        crate::GradesCommands::Course { course_id } => {
            let courses = get_enrolled_courses_api(&ctx.client, &ctx.session, "all").await?;
            let course_name = courses.iter()
                .find(|c| c.id == *course_id)
                .map(|c| c.fullname.as_str())
                .unwrap_or("");

            let mut grade = get_course_grades_api(&ctx.client, &ctx.session, *course_id, ctx.session.user_id).await?;
            grade.course_name = course_name.to_string();

            let mut rows: Vec<serde_json::Value> = Vec::new();
            rows.push(serde_json::json!({
                "courseId": grade.course_id,
                "courseName": &grade.course_name,
                "grade": grade.grade,
            }));

            for item in grade.items.as_deref().unwrap_or(&[]) {
                rows.push(serde_json::json!({
                    "name": item.name,
                    "grade": item.grade,
                    "percentage": item.percentage,
                    "weight": item.weight,
                    "feedback": item.feedback,
                    "graded": item.graded,
                }));
            }

            ctx.log.info(&format!("Found {} grade items", rows.len() - 1));
            format_and_output(&rows, ctx.output, None);
        }
    }

    Ok(())
}
